### PR TITLE
HybridWebView: Invoke .NET methods from JavaScript

### LINKED ISSUE #22304: HybridWebView: Add support for JavaScript to invoke C# methods
This is a placeholder for now. Will require design work.

### PR DESCRIPTION
**Pull request by Eilon**
### Description of Change

This change enables JavaScript code to invoke .NET methods with the HybridWebView control.

Example:

### 1. Define the C# methods that are callable

In the .NET code a class is declared that defines the methods that are callable from JavaScript:

```c#
private class DotNetMethods
{
	private HybridWebViewPage _hybridWebViewPage;

	public DotNetMethods(HybridWebViewPage hybridWebViewPage)
	{
		_hybridWebViewPage = hybridWebViewPage;
	}

	public void DoSyncWork()
	{
		Debug.WriteLine("DoSyncWork");
	}

	public void DoSyncWorkParams(int i, string s)
	{
		Debug.WriteLine($"DoSyncWorkParams: {i}, {s}");
	}

	public string DoSyncWorkReturn()
	{
		Debug.WriteLine("DoSyncWorkReturn");
		return "Hello from C#!";
	}

	public SyncReturn DoSyncWorkParamsReturn(int i, string s)
	{
		Debug.WriteLine($"DoSyncWorkParamsReturn: {i}, {s}");
		return new SyncReturn
		{
			Message = "Hello from C#! " + s,
			Value = i,
		};
	}
}
```

### 2. Tell the HybridWebView to use this type

```c#
hwv.InvokeJavaScriptTarget = new DotNetMethods(this);
```

### 3. Call the C# methods from JavaScript

```js
const retValue = await window.HybridWebView.InvokeDotNet('DoSyncWorkParamsReturn', [123, 'hello']);
```

### Notes

1. Though the JavaScript method is async, only sync .NET methods are supported. The technique used to call .NET methods for some platforms is sync-only, so there's no easy way to call async methods without blocking (which can deadlock). This will likely be the state of things for .NET 9.

### Issues Fixed

Fixes #22304

### CODE REVIEW

**Comment by egvijayanand** [comment]
Wow, about to log this request.

**Review by Redth — DISMISSED** [review]

**Comment by jonathanpeppers** [comment]
The current trimmer setup is quite a bit closer, the warnings we get now are from a single line of code:
```
No warnings expected, but got 2 warnings:
//src/Controls/src/Core/Hosting/AppHostBuilderExtensions.cs: IL2026: Microsoft.Maui.Controls.Hosting.AppHostBuilderExtensions.AddControlsHandlers(IMauiHandlersCollection): Using member 'Microsoft.Maui.Handlers.HybridWebViewHandler.HybridWebViewHandler()' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. HybridWebView uses dynamic System.Text.Json serialization features.
//src/Controls/src/Core/Hosting/AppHostBuilderExtensions.cs: IL2026: Microsoft.Maui.Controls.Hosting.AppHostBuilderExtensions.AddControlsHandlers(IMauiHandlersCollection): Using member 'Microsoft.Maui.Handlers.HybridWebViewHandler.HybridWebViewHandler(IPropertyMapper,CommandMapper)' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. HybridWebView uses dynamic System.Text.Json serialization features.
//src/Controls/src/Core/Hosting/AppHostBuilderExtensions.cs: IL2026: Microsoft.Maui.Controls.Hosting.AppHostBuilderExtensions.AddControlsHandlers(IMauiHandlersCollection): Using member 'Microsoft.Maui.Handlers.HybridWebViewHandler.HybridWebViewHandler()' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. HybridWebView uses dynamic System.Text.Json serialization features.
//src/Controls/src/Core/Hosting/AppHostBuilderExtensions.cs: IL3050: Microsoft.Maui.Controls.Hosting.AppHostBuilderExtensions.AddControlsHandlers(IMauiHandlersCollection): Using member 'Microsoft.Maui.Handlers.HybridWebViewHandler.HybridWebViewHandler()' which has 'RequiresDynamicCodeAttribute' can break functionality when AOT compiling. HybridWebView uses dynamic System.Text.Json serialization features.
//src/Controls/src/Core/Hosting/AppHostBuilderExtensions.cs: IL3050: Microsoft.Maui.Controls.Hosting.AppHostBuilderExtensions.AddControlsHandlers(IMauiHandlersCollection): Using member 'Microsoft.Maui.Handlers.HybridWebViewHandler.HybridWebViewHandler(IPropertyMapper,CommandMapper)' which has 'RequiresDynamicCodeAttribute' can break functionality when AOT compiling. HybridWebView uses dynamic System.Text.Json serialization features.
//src/Controls/src/Core/Hosting/AppHostBuilderExtensions.cs: IL3050: Microsoft.Maui.Controls.Hosting.AppHostBuilderExtensions.AddControlsHandlers(IMauiHandlersCollection): Using member 'Microsoft.Maui.Handlers.HybridWebViewHandler.HybridWebViewHandler()' which has 'RequiresDynamicCodeAttribute' can break functionality when AOT compiling. HybridWebView uses dynamic System.Text.Json serialization features.
/_/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/EnumConverterFactory.cs: IL3050: System.Text.Json.Serialization.Converters.EnumConverterFactory.CreateConverter(Type,JsonSerializerOptions): Using member 'System.Text.Json.Serialization.Converters.EnumConverterFactory.Create(Type,EnumConverterOptions,JsonNamingPolicy,JsonSerializerOptions)' which has 'RequiresDynamicCodeAttribute' can break functionality when AOT compiling. JSON serialization and deserialization might require types that cannot be statically analyzed and might need runtime code generation. Use System.Text.Json source generation for native AOT applications.
```

So, it seems like this isn't being inlined:
```csharp
if (RuntimeFeature.IsHybridWebViewSupported)
{
	// NOTE: not registered under NativeAOT or TrimMode=Full scenarios
	handlersCollection.AddHandler<HybridWebView, HybridWebViewHandler>();
}
```
And if we fix that, it should be 0 warnings.

File: src/Controls/src/Build.Tasks/nuget/buildTransitive/netstandard2.0/Microsoft.Maui.Controls.targets [hunk 1]
```diff
@@ -253,14 +254,18 @@
...
                                       Trim="true" />
-      <RuntimeHostConfigurationOption Include="Microsoft.Maui.RuntimeFeature.Microsoft.Maui.RuntimeFeature.IsXamlCBindingWithSourceCompilationEnabled"
+      <RuntimeHostConfigurationOption Include="Microsoft.Maui.RuntimeFeature.IsXamlCBindingWithSourceCompilationEnabled"
                                       Condition="'$(MauiEnableXamlCBindingWithSourceCompilation)' != ''"
                                       Value="$(MauiEnableXamlCBindingWithSourceCompilation)"
                                       Trim="false" />
+      <RuntimeHostConfigurationOption Include="Microsoft.Maui.RuntimeFeature.IsHybridWebViewSupported"
+                                      Condition="'$(MauiHybridWebViewSupported)' != ''"
+                                      Value="$(MauiHybridWebViewSupported)"
+                                      Trim="true" />
```
Comment (jonathanpeppers) [inline src/Controls/src/Build.Tasks/nuget/buildTransitive/netstandard2.0/Microsoft.Maui.Controls.targets:268]: These `@(RuntimeHostConfigurationOption)` items had a duplicate `Microsoft.Maui.RuntimeFeature.` prefix!

**Comment by jonathanpeppers** [comment]
The latest looks pretty good!

image

**Comment by rmarinho** [comment]
Since this is a new feature this should be 9.1.0 and not a 9.0.1 in theory right? 
So we should not merge this to main that is the sr1 ?!

**Comment by Eilon** [comment]
@Redth / @PureWeen - this is green on CI now (hugest thanks to @jonathanpeppers !). Is this good to go for 9.0.1xx?

File: src/Core/src/Handlers/HybridWebView/HybridWebViewHandler.cs [hunk 2]
```diff
@@ -116,22 +128,135 @@ void MessageReceived(string rawMessage)
...
+			public string[]? ParamValues { get; set; }
+		}
+
+		private sealed class DotNetInvokeResult
+		{
+			public object? Result { get; set; }
+			public bool IsJson { get; set; }
+		}
+
+		[JsonSourceGenerationOptions()]
```
Comment (jkurdek) [inline src/Core/src/Handlers/HybridWebView/HybridWebViewHandler.cs:252]: What does the empty `[JsonSourceGenerationOptions()]` do?

**Review by PureWeen — APPROVED** [review]

**Comment by jkurdek** [comment]
The `IL3050` warning issue is tracked in https://github.com/dotnet/runtime/issues/109177

**Pull request merged**